src/writer.rs

```diff
@@ -146,70 +179,304 @@ impl<'a, W: Write> Writer<'a, W> {
+/// schema validation.
+///
+/// This is an internal function which gets the bytes buffer where to write as parameter instead of
+/// creating a new one like `to_avro_datum`.
+fn write_avro_datum<T: ToAvro>(
+    schema: &Schema,
+    value: T,
+    buffer: &mut Vec<u8>,
+) -> Result<(), Error> {
+    let avro = value.avro();
+    if !avro.validate(schema) {
+        return Err(err_msg("value does not match schema"))
+    }
+    // TODO: this resolve call is super-expensive, try to pass the schema to avro() instead
+    Ok(encode(avro.resolve(schema)?, buffer))
+}
 
-        writer.write_all(self.header.as_slice())?;
-        writer.write_all(encode_raw(1i64.avro()).as_ref())?;
-        writer.write_all(encode_raw(num_bytes.avro()).as_ref())?;
-        writer.write_all(self.buffer.as_ref())?;
-        writer.write_all(SINGLE_WRITER_MARKER)?;
+/// Encode a compatible value (implementing the `ToAvro` trait) into Avro format, also
+/// performing schema validation.
+///
+/// **NOTE** This function has a quite small niche of usage and does NOT generate headers and sync
+/// markers; use [`Writer`](struct.Writer.html) to be fully Avro-compatible if you don't know what
+/// you are doing, instead.
+pub fn to_avro_datum<T: ToAvro>(schema: &Schema, value: T) -> Result<Vec<u8>, Error> {
+    let mut buffer = Vec::new();
+    write_avro_datum(schema, value, &mut buffer)?;
+    Ok(buffer)
+}
 
-        self.buffer.clear();
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use types::Record;
+    use util::zig_i64;
+
+    static SCHEMA: &'static str = r#"
+            {
+                "type": "record",
+                "name": "test",
+                "fields": [
+                    {"name": "a", "type": "long", "default": 42},
+                    {"name": "b", "type": "string"}
+                ]
+            }
+        "#;
+
+    #[test]
+    fn test_to_avro_datum() {
+        let schema = Schema::parse_str(SCHEMA).unwrap();
+        let mut record = Record::new(&schema).unwrap();
+        record.put("a", 27i64);
+        record.put("b", "foo");
+
+        let mut expected = Vec::new();
+        zig_i64(27, &mut expected);
+        zig_i64(3, &mut expected);
+        expected.extend(vec![b'f', b'o', b'o'].into_iter());
+
+        assert_eq!(to_avro_datum(&schema, record).unwrap(), expected);
+    }
 
-        Ok(())
+    #[test]
+    fn test_writer_append() {
+        let schema = Schema::parse_str(SCHEMA).unwrap();
+        let mut writer = Writer::new(&schema, Vec::new());
+
+        let mut record = Record::new(&schema).unwrap();
+        record.put("a", 27i64);
+        record.put("b", "foo");
+
+        let n1 = writer.append(record.clone()).unwrap();
+        let n2 = writer.append(record.clone()).unwrap();
+        let n3 = writer.flush().unwrap();
+        let result = writer.into_inner();
+
+        // TODO: Something is fishy here; the numbers don't add up
+        // assert_eq!(n1 + n2 + n3, result.len());
```

Comment:
Probably the mistake I did, mentioned in an earlier comment (number of bytes for the header not taken into account)?